// ibm-watson/text-to-speech/v1-generated.js
"use strict";
/**
 * (C) Copyright IBM Corp. 2017, 2020.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * IBM OpenAPI SDK Code Generator Version: 99-SNAPSHOT-be3b4618-20201221-123327
 */
var extend = require("extend");
var ibm_cloud_sdk_core_1 = require("ibm-cloud-sdk-core");
var common_1 = require("../lib/common");
/**
 * The IBM Watson&trade; Text to Speech service provides APIs that use IBM's speech-synthesis capabilities to synthesize
 * text into natural-sounding speech in a variety of languages, dialects, and voices. The service supports at least one
 * male or female voice, sometimes both, for each language. The audio is streamed back to the client with minimal delay.
 *
 *
 * For speech synthesis, the service supports a synchronous HTTP Representational State Transfer (REST) interface and a
 * WebSocket interface. Both interfaces support plain text and SSML input. SSML is an XML-based markup language that
 * provides text annotation for speech-synthesis applications. The WebSocket interface also supports the SSML
 * <code>&lt;mark&gt;</code> element and word timings.
 *
 * The service offers a customization interface that you can use to define sounds-like or phonetic translations for
 * words. A sounds-like translation consists of one or more words that, when combined, sound like the word. A phonetic
 * translation is based on the SSML phoneme format for representing a word. You can specify a phonetic translation in
 * standard International Phonetic Alphabet (IPA) representation or in the proprietary IBM Symbolic Phonetic
 * Representation (SPR). The Arabic, Chinese, Dutch, Australian English, and Korean languages support only IPA.
 */
var TextToSpeechV1 = /** @class */ (function (_super) {
    __extends(TextToSpeechV1, _super);
    /**
     * Construct a TextToSpeechV1 object.
     *
     * @param {Object} options - Options for the service.
     * @param {string} [options.serviceUrl] - The base url to use when contacting the service. The base url may differ between IBM Cloud regions.
     * @param {OutgoingHttpHeaders} [options.headers] - Default headers that shall be included with every request to the service.
     * @param {string} [options.serviceName] - The name of the service to configure
     * @param {Authenticator} [options.authenticator] - The Authenticator object used to authenticate requests to the service. Defaults to environment if not set
     * @constructor
     * @returns {TextToSpeechV1}
     */
    function TextToSpeechV1(options) {
        var _this = this;
        options = options || {};
        if (!options.serviceName) {
            options.serviceName = TextToSpeechV1.DEFAULT_SERVICE_NAME;
        }
        // If the caller didn't supply an authenticator, construct one from external configuration.
        if (!options.authenticator) {
            options.authenticator = ibm_cloud_sdk_core_1.getAuthenticatorFromEnvironment(options.serviceName);
        }
        _this = _super.call(this, options) || this;
        _this.configureService(options.serviceName);
        if (options.serviceUrl) {
            _this.setServiceUrl(options.serviceUrl);
        }
        return _this;
    }
    /*************************
     * voices
     ************************/
    /**
     * List voices.
     *
     * Lists all voices available for use with the service. The information includes the name, language, gender, and other
     * details about the voice. The ordering of the list of voices can change from call to call; do not rely on an
     * alphabetized or static list of voices. To see information about a specific voice, use the **Get a voice** method.
     *
     * **See also:** [Listing all available
     * voices](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-voices#listVoices).
     *
     * @param {Object} [params] - The parameters to send to the service.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Voices>>}
     */
    TextToSpeechV1.prototype.listVoices = function (params) {
        var _params = Object.assign({}, params);
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'listVoices');
        var parameters = {
            options: {
                url: '/v1/voices',
                method: 'GET',
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Get a voice.
     *
     * Gets information about the specified voice. The information includes the name, language, gender, and other details
     * about the voice. Specify a customization ID to obtain information for a custom model that is defined for the
     * language of the specified voice. To list information about all available voices, use the **List voices** method.
     *
     * **See also:** [Listing a specific
     * voice](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-voices#listVoice).
     *
     * ### Important voice updates
     *
     *  The service's voices underwent significant change on 2 December 2020.
     * * The Arabic, Chinese, Dutch, Australian English, and Korean voices are now neural instead of concatenative.
     * * The `ar-AR_OmarVoice` voice is deprecated. Use `ar-MS_OmarVoice` voice instead.
     * * The `ar-AR` language identifier cannot be used to create a custom model. Use the `ar-MS` identifier instead.
     * * The standard concatenative voices for the following languages are now deprecated: Brazilian Portuguese, United
     * Kingdom and United States English, French, German, Italian, Japanese, and Spanish (all dialects).
     * * The features expressive SSML, voice transformation SSML, and use of the `volume` attribute of the `<prosody>`
     * element are deprecated and are not supported with any of the service's neural voices.
     * * All of the service's voices are now customizable and generally available (GA) for production use.
     *
     * The deprecated voices and features will continue to function for at least one year but might be removed at a future
     * date. You are encouraged to migrate to the equivalent neural voices at your earliest convenience. For more
     * information about all voice updates, see the [2 December 2020 service
     * update](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-release-notes#December2020) in the release
     * notes.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.voice - The voice for which information is to be returned. For more information about
     * specifying a voice, see **Important voice updates** in the method description.
     * @param {string} [params.customizationId] - The customization ID (GUID) of a custom model for which information is
     * to be returned. You must make the request with credentials for the instance of the service that owns the custom
     * model. Omit the parameter to see information about the specified voice with no customization.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Voice>>}
     */
    TextToSpeechV1.prototype.getVoice = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['voice'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var query = {
            'customization_id': _params.customizationId
        };
        var path = {
            'voice': _params.voice
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'getVoice');
        var parameters = {
            options: {
                url: '/v1/voices/{voice}',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /*************************
     * synthesis
     ************************/
    /**
     * Synthesize audio.
     *
     * Synthesizes text to audio that is spoken in the specified voice. The service bases its understanding of the
     * language for the input text on the specified voice. Use a voice that matches the language of the input text.
     *
     * The method accepts a maximum of 5 KB of input text in the body of the request, and 8 KB for the URL and headers.
     * The 5 KB limit includes any SSML tags that you specify. The service returns the synthesized audio stream as an
     * array of bytes.
     *
     * **See also:** [The HTTP
     * interface](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-usingHTTP#usingHTTP).
     *
     * ### Audio formats (accept types)
     *
     *  The service can return audio in the following formats (MIME types).
     * * Where indicated, you can optionally specify the sampling rate (`rate`) of the audio. You must specify a sampling
     * rate for the `audio/l16` and `audio/mulaw` formats. A specified sampling rate must lie in the range of 8 kHz to 192
     * kHz. Some formats restrict the sampling rate to certain values, as noted.
     * * For the `audio/l16` format, you can optionally specify the endianness (`endianness`) of the audio:
     * `endianness=big-endian` or `endianness=little-endian`.
     *
     * Use the `Accept` header or the `accept` parameter to specify the requested format of the response audio. If you
     * omit an audio format altogether, the service returns the audio in Ogg format with the Opus codec
     * (`audio/ogg;codecs=opus`). The service always returns single-channel audio.
     * * `audio/basic` - The service returns audio with a sampling rate of 8000 Hz.
     * * `audio/flac` - You can optionally specify the `rate` of the audio. The default sampling rate is 22,050 Hz.
     * * `audio/l16` - You must specify the `rate` of the audio. You can optionally specify the `endianness` of the audio.
     * The default endianness is `little-endian`.
     * * `audio/mp3` - You can optionally specify the `rate` of the audio. The default sampling rate is 22,050 Hz.
     * * `audio/mpeg` - You can optionally specify the `rate` of the audio. The default sampling rate is 22,050 Hz.
     * * `audio/mulaw` - You must specify the `rate` of the audio.
     * * `audio/ogg` - The service returns the audio in the `vorbis` codec. You can optionally specify the `rate` of the
     * audio. The default sampling rate is 22,050 Hz.
     * * `audio/ogg;codecs=opus` - You can optionally specify the `rate` of the audio. Only the following values are valid
     * sampling rates: `48000`, `24000`, `16000`, `12000`, or `8000`. If you specify a value other than one of these, the
     * service returns an error. The default sampling rate is 48,000 Hz.
     * * `audio/ogg;codecs=vorbis` - You can optionally specify the `rate` of the audio. The default sampling rate is
     * 22,050 Hz.
     * * `audio/wav` - You can optionally specify the `rate` of the audio. The default sampling rate is 22,050 Hz.
     * * `audio/webm` - The service returns the audio in the `opus` codec. The service returns audio with a sampling rate
     * of 48,000 Hz.
     * * `audio/webm;codecs=opus` - The service returns audio with a sampling rate of 48,000 Hz.
     * * `audio/webm;codecs=vorbis` - You can optionally specify the `rate` of the audio. The default sampling rate is
     * 22,050 Hz.
     *
     * For more information about specifying an audio format, including additional details about some of the formats, see
     * [Audio formats](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-audioFormats#audioFormats).
     *
     * ### Important voice updates
     *
     *  The service's voices underwent significant change on 2 December 2020.
     * * The Arabic, Chinese, Dutch, Australian English, and Korean voices are now neural instead of concatenative.
     * * The `ar-AR_OmarVoice` voice is deprecated. Use `ar-MS_OmarVoice` voice instead.
     * * The `ar-AR` language identifier cannot be used to create a custom model. Use the `ar-MS` identifier instead.
     * * The standard concatenative voices for the following languages are now deprecated: Brazilian Portuguese, United
     * Kingdom and United States English, French, German, Italian, Japanese, and Spanish (all dialects).
     * * The features expressive SSML, voice transformation SSML, and use of the `volume` attribute of the `<prosody>`
     * element are deprecated and are not supported with any of the service's neural voices.
     * * All of the service's voices are now customizable and generally available (GA) for production use.
     *
     * The deprecated voices and features will continue to function for at least one year but might be removed at a future
     * date. You are encouraged to migrate to the equivalent neural voices at your earliest convenience. For more
     * information about all voice updates, see the [2 December 2020 service
     * update](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-release-notes#December2020) in the release
     * notes.
     *
     * ### Warning messages
     *
     *  If a request includes invalid query parameters, the service returns a `Warnings` response header that provides
     * messages about the invalid parameters. The warning includes a descriptive message and a list of invalid argument
     * strings. For example, a message such as `"Unknown arguments:"` or `"Unknown url query arguments:"` followed by a
     * list of the form `"{invalid_arg_1}, {invalid_arg_2}."` The request succeeds despite the warnings.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.text - The text to synthesize.
     * @param {string} [params.accept] - The requested format (MIME type) of the audio. You can use the `Accept` header or
     * the `accept` parameter to specify the audio format. For more information about specifying an audio format, see
     * **Audio formats (accept types)** in the method description.
     * @param {string} [params.voice] - The voice to use for synthesis. For more information about specifying a voice, see
     * **Important voice updates** in the method description.
     * @param {string} [params.customizationId] - The customization ID (GUID) of a custom model to use for the synthesis.
     * If a custom model is specified, it works only if it matches the language of the indicated voice. You must make the
     * request with credentials for the instance of the service that owns the custom model. Omit the parameter to use the
     * specified voice with no customization.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<NodeJS.ReadableStream|Buffer>>}
     */
    TextToSpeechV1.prototype.synthesize = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['text'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var body = {
            'text': _params.text
        };
        var query = {
            'voice': _params.voice,
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'synthesize');
        var parameters = {
            options: {
                url: '/v1/synthesize',
                method: 'POST',
                body: body,
                qs: query,
                responseType: 'stream',
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Content-Type': 'application/json',
                    'Accept': _params.accept
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /*************************
     * pronunciation
     ************************/
    /**
     * Get pronunciation.
     *
     * Gets the phonetic pronunciation for the specified word. You can request the pronunciation for a specific format.
     * You can also request the pronunciation for a specific voice to see the default translation for the language of that
     * voice or for a specific custom model to see the translation for that model.
     *
     * **See also:** [Querying a word from a
     * language](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuWordsQueryLanguage).
     *
     * ### Important voice updates
     *
     *  The service's voices underwent significant change on 2 December 2020.
     * * The Arabic, Chinese, Dutch, Australian English, and Korean voices are now neural instead of concatenative.
     * * The `ar-AR_OmarVoice` voice is deprecated. Use `ar-MS_OmarVoice` voice instead.
     * * The `ar-AR` language identifier cannot be used to create a custom model. Use the `ar-MS` identifier instead.
     * * The standard concatenative voices for the following languages are now deprecated: Brazilian Portuguese, United
     * Kingdom and United States English, French, German, Italian, Japanese, and Spanish (all dialects).
     * * The features expressive SSML, voice transformation SSML, and use of the `volume` attribute of the `<prosody>`
     * element are deprecated and are not supported with any of the service's neural voices.
     * * All of the service's voices are now customizable and generally available (GA) for production use.
     *
     * The deprecated voices and features will continue to function for at least one year but might be removed at a future
     * date. You are encouraged to migrate to the equivalent neural voices at your earliest convenience. For more
     * information about all voice updates, see the [2 December 2020 service
     * update](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-release-notes#December2020) in the release
     * notes.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.text - The word for which the pronunciation is requested.
     * @param {string} [params.voice] - A voice that specifies the language in which the pronunciation is to be returned.
     * All voices for the same language (for example, `en-US`) return the same translation. For more information about
     * specifying a voice, see **Important voice updates** in the method description.
     * @param {string} [params.format] - The phoneme format in which to return the pronunciation. The Arabic, Chinese,
     * Dutch, Australian English, and Korean languages support only IPA. Omit the parameter to obtain the pronunciation in
     * the default format.
     * @param {string} [params.customizationId] - The customization ID (GUID) of a custom model for which the
     * pronunciation is to be returned. The language of a specified custom model must match the language of the specified
     * voice. If the word is not defined in the specified custom model, the service returns the default translation for
     * the custom model's language. You must make the request with credentials for the instance of the service that owns
     * the custom model. Omit the parameter to see the translation for the specified voice with no customization.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Pronunciation>>}
     */
    TextToSpeechV1.prototype.getPronunciation = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['text'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var query = {
            'text': _params.text,
            'voice': _params.voice,
            'format': _params.format,
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'getPronunciation');
        var parameters = {
            options: {
                url: '/v1/pronunciation',
                method: 'GET',
                qs: query,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /*************************
     * customModels
     ************************/
    /**
     * Create a custom model.
     *
     * Creates a new empty custom model. You must specify a name for the new custom model. You can optionally specify the
     * language and a description for the new model. The model is owned by the instance of the service whose credentials
     * are used to create it.
     *
     * **See also:** [Creating a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customModels#cuModelsCreate).
     *
     * ### Important voice updates
     *
     *  The service's voices underwent significant change on 2 December 2020.
     * * The Arabic, Chinese, Dutch, Australian English, and Korean voices are now neural instead of concatenative.
     * * The `ar-AR_OmarVoice` voice is deprecated. Use `ar-MS_OmarVoice` voice instead.
     * * The `ar-AR` language identifier cannot be used to create a custom model. Use the `ar-MS` identifier instead.
     * * The standard concatenative voices for the following languages are now deprecated: Brazilian Portuguese, United
     * Kingdom and United States English, French, German, Italian, Japanese, and Spanish (all dialects).
     * * The features expressive SSML, voice transformation SSML, and use of the `volume` attribute of the `<prosody>`
     * element are deprecated and are not supported with any of the service's neural voices.
     * * All of the service's voices are now customizable and generally available (GA) for production use.
     *
     * The deprecated voices and features will continue to function for at least one year but might be removed at a future
     * date. You are encouraged to migrate to the equivalent neural voices at your earliest convenience. For more
     * information about all voice updates, see the [2 December 2020 service
     * update](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-release-notes#December2020) in the release
     * notes.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.name - The name of the new custom model.
     * @param {string} [params.language] - The language of the new custom model. You create a custom model for a specific
     * language, not for a specific voice. A custom model can be used with any voice for its specified language. Omit the
     * parameter to use the the default language, `en-US`. **Note:** The `ar-AR` language identifier cannot be used to
     * create a custom model. Use the `ar-MS` identifier instead.
     * @param {string} [params.description] - A description of the new custom model. Specifying a description is
     * recommended.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.CustomModel>>}
     */
    TextToSpeechV1.prototype.createCustomModel = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['name'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var body = {
            'name': _params.name,
            'language': _params.language,
            'description': _params.description
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'createCustomModel');
        var parameters = {
            options: {
                url: '/v1/customizations',
                method: 'POST',
                body: body,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * List custom models.
     *
     * Lists metadata such as the name and description for all custom models that are owned by an instance of the service.
     * Specify a language to list the custom models for that language only. To see the words in addition to the metadata
     * for a specific custom model, use the **List a custom model** method. You must use credentials for the instance of
     * the service that owns a model to list information about it.
     *
     * **See also:** [Querying all custom
     * models](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customModels#cuModelsQueryAll).
     *
     * @param {Object} [params] - The parameters to send to the service.
     * @param {string} [params.language] - The language for which custom models that are owned by the requesting
     * credentials are to be returned. Omit the parameter to see all custom models that are owned by the requester.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.CustomModels>>}
     */
    TextToSpeechV1.prototype.listCustomModels = function (params) {
        var _params = Object.assign({}, params);
        var query = {
            'language': _params.language
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'listCustomModels');
        var parameters = {
            options: {
                url: '/v1/customizations',
                method: 'GET',
                qs: query,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Update a custom model.
     *
     * Updates information for the specified custom model. You can update metadata such as the name and description of the
     * model. You can also update the words in the model and their translations. Adding a new translation for a word that
     * already exists in a custom model overwrites the word's existing translation. A custom model can contain no more
     * than 20,000 entries. You must use credentials for the instance of the service that owns a model to update it.
     *
     * You can define sounds-like or phonetic translations for words. A sounds-like translation consists of one or more
     * words that, when combined, sound like the word. Phonetic translations are based on the SSML phoneme format for
     * representing a word. You can specify them in standard International Phonetic Alphabet (IPA) representation
     *
     *   <code>&lt;phoneme alphabet="ipa" ph="t&#601;m&#712;&#593;to"&gt;&lt;/phoneme&gt;</code>
     *
     *   or in the proprietary IBM Symbolic Phonetic Representation (SPR)
     *
     *   <code>&lt;phoneme alphabet="ibm" ph="1gAstroEntxrYFXs"&gt;&lt;/phoneme&gt;</code>
     *
     * **See also:**
     * * [Updating a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customModels#cuModelsUpdate)
     * * [Adding words to a Japanese custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuJapaneseAdd)
     * * [Understanding
     * customization](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customIntro#customIntro).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {string} [params.name] - A new name for the custom model.
     * @param {string} [params.description] - A new description for the custom model.
     * @param {Word[]} [params.words] - An array of `Word` objects that provides the words and their translations that are
     * to be added or updated for the custom model. Pass an empty array to make no additions or updates.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Empty>>}
     */
    TextToSpeechV1.prototype.updateCustomModel = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'words': _params.words
        };
        var path = {
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'updateCustomModel');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Get a custom model.
     *
     * Gets all information about a specified custom model. In addition to metadata such as the name and description of
     * the custom model, the output includes the words and their translations as defined in the model. To see just the
     * metadata for a model, use the **List custom models** method.
     *
     * **See also:** [Querying a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customModels#cuModelsQuery).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.CustomModel>>}
     */
    TextToSpeechV1.prototype.getCustomModel = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var path = {
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'getCustomModel');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Delete a custom model.
     *
     * Deletes the specified custom model. You must use credentials for the instance of the service that owns a model to
     * delete it.
     *
     * **See also:** [Deleting a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customModels#cuModelsDelete).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Empty>>}
     */
    TextToSpeechV1.prototype.deleteCustomModel = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var path = {
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteCustomModel');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /*************************
     * customWords
     ************************/
    /**
     * Add custom words.
     *
     * Adds one or more words and their translations to the specified custom model. Adding a new translation for a word
     * that already exists in a custom model overwrites the word's existing translation. A custom model can contain no
     * more than 20,000 entries. You must use credentials for the instance of the service that owns a model to add words
     * to it.
     *
     * You can define sounds-like or phonetic translations for words. A sounds-like translation consists of one or more
     * words that, when combined, sound like the word. Phonetic translations are based on the SSML phoneme format for
     * representing a word. You can specify them in standard International Phonetic Alphabet (IPA) representation
     *
     *   <code>&lt;phoneme alphabet="ipa" ph="t&#601;m&#712;&#593;to"&gt;&lt;/phoneme&gt;</code>
     *
     *   or in the proprietary IBM Symbolic Phonetic Representation (SPR)
     *
     *   <code>&lt;phoneme alphabet="ibm" ph="1gAstroEntxrYFXs"&gt;&lt;/phoneme&gt;</code>
     *
     * **See also:**
     * * [Adding multiple words to a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuWordsAdd)
     * * [Adding words to a Japanese custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuJapaneseAdd)
     * * [Understanding
     * customization](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customIntro#customIntro).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {Word[]} params.words - The **Add custom words** method accepts an array of `Word` objects. Each object
     * provides a word that is to be added or updated for the custom model and the word's translation.
     *
     * The **List custom words** method returns an array of `Word` objects. Each object shows a word and its translation
     * from the custom model. The words are listed in alphabetical order, with uppercase letters listed before lowercase
     * letters. The array is empty if the custom model contains no words.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Empty>>}
     */
    TextToSpeechV1.prototype.addWords = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId', 'words'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var body = {
            'words': _params.words
        };
        var path = {
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'addWords');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}/words',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * List custom words.
     *
     * Lists all of the words and their translations for the specified custom model. The output shows the translations as
     * they are defined in the model. You must use credentials for the instance of the service that owns a model to list
     * its words.
     *
     * **See also:** [Querying all words from a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuWordsQueryModel).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Words>>}
     */
    TextToSpeechV1.prototype.listWords = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var path = {
            'customization_id': _params.customizationId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'listWords');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}/words',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Add a custom word.
     *
     * Adds a single word and its translation to the specified custom model. Adding a new translation for a word that
     * already exists in a custom model overwrites the word's existing translation. A custom model can contain no more
     * than 20,000 entries. You must use credentials for the instance of the service that owns a model to add a word to
     * it.
     *
     * You can define sounds-like or phonetic translations for words. A sounds-like translation consists of one or more
     * words that, when combined, sound like the word. Phonetic translations are based on the SSML phoneme format for
     * representing a word. You can specify them in standard International Phonetic Alphabet (IPA) representation
     *
     *   <code>&lt;phoneme alphabet="ipa" ph="t&#601;m&#712;&#593;to"&gt;&lt;/phoneme&gt;</code>
     *
     *   or in the proprietary IBM Symbolic Phonetic Representation (SPR)
     *
     *   <code>&lt;phoneme alphabet="ibm" ph="1gAstroEntxrYFXs"&gt;&lt;/phoneme&gt;</code>
     *
     * **See also:**
     * * [Adding a single word to a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuWordAdd)
     * * [Adding words to a Japanese custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuJapaneseAdd)
     * * [Understanding
     * customization](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customIntro#customIntro).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {string} params.word - The word that is to be added or updated for the custom model.
     * @param {string} params.translation - The phonetic or sounds-like translation for the word. A phonetic translation
     * is based on the SSML format for representing the phonetic string of a word either as an IPA translation or as an
     * IBM SPR translation. The Arabic, Chinese, Dutch, Australian English, and Korean languages support only IPA. A
     * sounds-like is one or more words that, when combined, sound like the word.
     * @param {string} [params.partOfSpeech] - **Japanese only.** The part of speech for the word. The service uses the
     * value to produce the correct intonation for the word. You can create only a single entry, with or without a single
     * part of speech, for any word; you cannot create multiple entries with different parts of speech for the same word.
     * For more information, see [Working with Japanese
     * entries](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-rules#jaNotes).
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Empty>>}
     */
    TextToSpeechV1.prototype.addWord = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId', 'word', 'translation'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var body = {
            'translation': _params.translation,
            'part_of_speech': _params.partOfSpeech
        };
        var path = {
            'customization_id': _params.customizationId,
            'word': _params.word
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'addWord');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}/words/{word}',
                method: 'PUT',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Get a custom word.
     *
     * Gets the translation for a single word from the specified custom model. The output shows the translation as it is
     * defined in the model. You must use credentials for the instance of the service that owns a model to list its words.
     *
     *
     * **See also:** [Querying a single word from a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuWordQueryModel).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {string} params.word - The word that is to be queried from the custom model.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Translation>>}
     */
    TextToSpeechV1.prototype.getWord = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId', 'word'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var path = {
            'customization_id': _params.customizationId,
            'word': _params.word
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'getWord');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}/words/{word}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /**
     * Delete a custom word.
     *
     * Deletes a single word from the specified custom model. You must use credentials for the instance of the service
     * that owns a model to delete its words.
     *
     * **See also:** [Deleting a word from a custom
     * model](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-customWords#cuWordDelete).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customizationId - The customization ID (GUID) of the custom model. You must make the request
     * with credentials for the instance of the service that owns the custom model.
     * @param {string} params.word - The word that is to be deleted from the custom model.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Empty>>}
     */
    TextToSpeechV1.prototype.deleteWord = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customizationId', 'word'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var path = {
            'customization_id': _params.customizationId,
            'word': _params.word
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteWord');
        var parameters = {
            options: {
                url: '/v1/customizations/{customization_id}/words/{word}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    /*************************
     * userData
     ************************/
    /**
     * Delete labeled data.
     *
     * Deletes all data that is associated with a specified customer ID. The method deletes all data for the customer ID,
     * regardless of the method by which the information was added. The method has no effect if no data is associated with
     * the customer ID. You must issue the request with credentials for the same instance of the service that was used to
     * associate the customer ID with the data. You associate a customer ID with data by passing the `X-Watson-Metadata`
     * header with a request that passes the data.
     *
     * **Note:** If you delete an instance of the service from the service console, all data associated with that service
     * instance is automatically deleted. This includes all custom models and word/translation pairs, and all data related
     * to speech synthesis requests.
     *
     * **See also:** [Information
     * security](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-information-security#information-security).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.customerId - The customer ID for which all data is to be deleted.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<TextToSpeechV1.Response<TextToSpeechV1.Empty>>}
     */
    TextToSpeechV1.prototype.deleteUserData = function (params) {
        var _params = Object.assign({}, params);
        var requiredParams = ['customerId'];
        var missingParams = ibm_cloud_sdk_core_1.getMissingParams(_params, requiredParams);
        if (missingParams) {
            return Promise.reject(missingParams);
        }
        var query = {
            'customer_id': _params.customerId
        };
        var sdkHeaders = common_1.getSdkHeaders(TextToSpeechV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteUserData');
        var parameters = {
            options: {
                url: '/v1/user_data',
                method: 'DELETE',
                qs: query,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    ;
    TextToSpeechV1.DEFAULT_SERVICE_URL = 'https://api.us-south.text-to-speech.watson.cloud.ibm.com';
    TextToSpeechV1.DEFAULT_SERVICE_NAME = 'text_to_speech';
    return TextToSpeechV1;
}(ibm_cloud_sdk_core_1.BaseService));
/*************************
 * interfaces
 ************************/
(function (TextToSpeechV1) {
    /** Constants for the `getVoice` operation. */
    var GetVoiceConstants;
    (function (GetVoiceConstants) {
        /** The voice for which information is to be returned. For more information about specifying a voice, see **Important voice updates** in the method description. */
        var Voice;
        (function (Voice) {
            Voice["AR_AR_OMARVOICE"] = "ar-AR_OmarVoice";
            Voice["AR_MS_OMARVOICE"] = "ar-MS_OmarVoice";
            Voice["DE_DE_BIRGITVOICE"] = "de-DE_BirgitVoice";
            Voice["DE_DE_BIRGITV3VOICE"] = "de-DE_BirgitV3Voice";
            Voice["DE_DE_DIETERVOICE"] = "de-DE_DieterVoice";
            Voice["DE_DE_DIETERV3VOICE"] = "de-DE_DieterV3Voice";
            Voice["DE_DE_ERIKAV3VOICE"] = "de-DE_ErikaV3Voice";
            Voice["EN_AU_CRAIGVOICE"] = "en-AU-CraigVoice";
            Voice["EN_AU_MADISONVOICE"] = "en-AU-MadisonVoice";
            Voice["EN_GB_CHARLOTTEV3VOICE"] = "en-GB_CharlotteV3Voice";
            Voice["EN_GB_JAMESV3VOICE"] = "en-GB_JamesV3Voice";
            Voice["EN_GB_KATEVOICE"] = "en-GB_KateVoice";
            Voice["EN_GB_KATEV3VOICE"] = "en-GB_KateV3Voice";
            Voice["EN_US_ALLISONVOICE"] = "en-US_AllisonVoice";
            Voice["EN_US_ALLISONV3VOICE"] = "en-US_AllisonV3Voice";
            Voice["EN_US_EMILYV3VOICE"] = "en-US_EmilyV3Voice";
            Voice["EN_US_HENRYV3VOICE"] = "en-US_HenryV3Voice";
            Voice["EN_US_KEVINV3VOICE"] = "en-US_KevinV3Voice";
            Voice["EN_US_LISAVOICE"] = "en-US_LisaVoice";
            Voice["EN_US_LISAV3VOICE"] = "en-US_LisaV3Voice";
            Voice["EN_US_MICHAELVOICE"] = "en-US_MichaelVoice";
            Voice["EN_US_MICHAELV3VOICE"] = "en-US_MichaelV3Voice";
            Voice["EN_US_OLIVIAV3VOICE"] = "en-US_OliviaV3Voice";
            Voice["ES_ES_ENRIQUEVOICE"] = "es-ES_EnriqueVoice";
            Voice["ES_ES_ENRIQUEV3VOICE"] = "es-ES_EnriqueV3Voice";
            Voice["ES_ES_LAURAVOICE"] = "es-ES_LauraVoice";
            Voice["ES_ES_LAURAV3VOICE"] = "es-ES_LauraV3Voice";
            Voice["ES_LA_SOFIAVOICE"] = "es-LA_SofiaVoice";
            Voice["ES_LA_SOFIAV3VOICE"] = "es-LA_SofiaV3Voice";
            Voice["ES_US_SOFIAVOICE"] = "es-US_SofiaVoice";
            Voice["ES_US_SOFIAV3VOICE"] = "es-US_SofiaV3Voice";
            Voice["FR_FR_NICOLASV3VOICE"] = "fr-FR_NicolasV3Voice";
            Voice["FR_FR_RENEEVOICE"] = "fr-FR_ReneeVoice";
            Voice["FR_FR_RENEEV3VOICE"] = "fr-FR_ReneeV3Voice";
            Voice["IT_IT_FRANCESCAVOICE"] = "it-IT_FrancescaVoice";
            Voice["IT_IT_FRANCESCAV3VOICE"] = "it-IT_FrancescaV3Voice";
            Voice["JA_JP_EMIVOICE"] = "ja-JP_EmiVoice";
            Voice["JA_JP_EMIV3VOICE"] = "ja-JP_EmiV3Voice";
            Voice["KO_KR_HYUNJUNVOICE"] = "ko-KR_HyunjunVoice";
            Voice["KO_KR_SIWOOVOICE"] = "ko-KR_SiWooVoice";
            Voice["KO_KR_YOUNGMIVOICE"] = "ko-KR_YoungmiVoice";
            Voice["KO_KR_YUNAVOICE"] = "ko-KR_YunaVoice";
            Voice["NL_NL_EMMAVOICE"] = "nl-NL_EmmaVoice";
            Voice["NL_NL_LIAMVOICE"] = "nl-NL_LiamVoice";
            Voice["PT_BR_ISABELAVOICE"] = "pt-BR_IsabelaVoice";
            Voice["PT_BR_ISABELAV3VOICE"] = "pt-BR_IsabelaV3Voice";
            Voice["ZH_CN_LINAVOICE"] = "zh-CN_LiNaVoice";
            Voice["ZH_CN_WANGWEIVOICE"] = "zh-CN_WangWeiVoice";
            Voice["ZH_CN_ZHANGJINGVOICE"] = "zh-CN_ZhangJingVoice";
        })(Voice = GetVoiceConstants.Voice || (GetVoiceConstants.Voice = {}));
    })(GetVoiceConstants = TextToSpeechV1.GetVoiceConstants || (TextToSpeechV1.GetVoiceConstants = {}));
    /** Constants for the `synthesize` operation. */
    var SynthesizeConstants;
    (function (SynthesizeConstants) {
        /** The requested format (MIME type) of the audio. You can use the `Accept` header or the `accept` parameter to specify the audio format. For more information about specifying an audio format, see **Audio formats (accept types)** in the method description. */
        var Accept;
        (function (Accept) {
            Accept["AUDIO_BASIC"] = "audio/basic";
            Accept["AUDIO_FLAC"] = "audio/flac";
            Accept["AUDIO_L16"] = "audio/l16";
            Accept["AUDIO_OGG"] = "audio/ogg";
            Accept["AUDIO_OGG_CODECS_OPUS"] = "audio/ogg;codecs=opus";
            Accept["AUDIO_OGG_CODECS_VORBIS"] = "audio/ogg;codecs=vorbis";
            Accept["AUDIO_MP3"] = "audio/mp3";
            Accept["AUDIO_MPEG"] = "audio/mpeg";
            Accept["AUDIO_MULAW"] = "audio/mulaw";
            Accept["AUDIO_WAV"] = "audio/wav";
            Accept["AUDIO_WEBM"] = "audio/webm";
            Accept["AUDIO_WEBM_CODECS_OPUS"] = "audio/webm;codecs=opus";
            Accept["AUDIO_WEBM_CODECS_VORBIS"] = "audio/webm;codecs=vorbis";
        })(Accept = SynthesizeConstants.Accept || (SynthesizeConstants.Accept = {}));
        /** The voice to use for synthesis. For more information about specifying a voice, see **Important voice updates** in the method description. */
        var Voice;
        (function (Voice) {
            Voice["AR_AR_OMARVOICE"] = "ar-AR_OmarVoice";
            Voice["AR_MS_OMARVOICE"] = "ar-MS_OmarVoice";
            Voice["DE_DE_BIRGITVOICE"] = "de-DE_BirgitVoice";
            Voice["DE_DE_BIRGITV3VOICE"] = "de-DE_BirgitV3Voice";
            Voice["DE_DE_DIETERVOICE"] = "de-DE_DieterVoice";
            Voice["DE_DE_DIETERV3VOICE"] = "de-DE_DieterV3Voice";
            Voice["DE_DE_ERIKAV3VOICE"] = "de-DE_ErikaV3Voice";
            Voice["EN_AU_CRAIGVOICE"] = "en-AU-CraigVoice";
            Voice["EN_AU_MADISONVOICE"] = "en-AU-MadisonVoice";
            Voice["EN_GB_CHARLOTTEV3VOICE"] = "en-GB_CharlotteV3Voice";
            Voice["EN_GB_JAMESV3VOICE"] = "en-GB_JamesV3Voice";
            Voice["EN_GB_KATEVOICE"] = "en-GB_KateVoice";
            Voice["EN_GB_KATEV3VOICE"] = "en-GB_KateV3Voice";
            Voice["EN_US_ALLISONVOICE"] = "en-US_AllisonVoice";
            Voice["EN_US_ALLISONV3VOICE"] = "en-US_AllisonV3Voice";
            Voice["EN_US_EMILYV3VOICE"] = "en-US_EmilyV3Voice";
            Voice["EN_US_HENRYV3VOICE"] = "en-US_HenryV3Voice";
            Voice["EN_US_KEVINV3VOICE"] = "en-US_KevinV3Voice";
            Voice["EN_US_LISAVOICE"] = "en-US_LisaVoice";
            Voice["EN_US_LISAV3VOICE"] = "en-US_LisaV3Voice";
            Voice["EN_US_MICHAELVOICE"] = "en-US_MichaelVoice";
            Voice["EN_US_MICHAELV3VOICE"] = "en-US_MichaelV3Voice";
            Voice["EN_US_OLIVIAV3VOICE"] = "en-US_OliviaV3Voice";
            Voice["ES_ES_ENRIQUEVOICE"] = "es-ES_EnriqueVoice";
            Voice["ES_ES_ENRIQUEV3VOICE"] = "es-ES_EnriqueV3Voice";
            Voice["ES_ES_LAURAVOICE"] = "es-ES_LauraVoice";
            Voice["ES_ES_LAURAV3VOICE"] = "es-ES_LauraV3Voice";
            Voice["ES_LA_SOFIAVOICE"] = "es-LA_SofiaVoice";
            Voice["ES_LA_SOFIAV3VOICE"] = "es-LA_SofiaV3Voice";
            Voice["ES_US_SOFIAVOICE"] = "es-US_SofiaVoice";
            Voice["ES_US_SOFIAV3VOICE"] = "es-US_SofiaV3Voice";
            Voice["FR_FR_NICOLASV3VOICE"] = "fr-FR_NicolasV3Voice";
            Voice["FR_FR_RENEEVOICE"] = "fr-FR_ReneeVoice";
            Voice["FR_FR_RENEEV3VOICE"] = "fr-FR_ReneeV3Voice";
            Voice["IT_IT_FRANCESCAVOICE"] = "it-IT_FrancescaVoice";
            Voice["IT_IT_FRANCESCAV3VOICE"] = "it-IT_FrancescaV3Voice";
            Voice["JA_JP_EMIVOICE"] = "ja-JP_EmiVoice";
            Voice["JA_JP_EMIV3VOICE"] = "ja-JP_EmiV3Voice";
            Voice["KO_KR_HYUNJUNVOICE"] = "ko-KR_HyunjunVoice";
            Voice["KO_KR_SIWOOVOICE"] = "ko-KR_SiWooVoice";
            Voice["KO_KR_YOUNGMIVOICE"] = "ko-KR_YoungmiVoice";
            Voice["KO_KR_YUNAVOICE"] = "ko-KR_YunaVoice";
            Voice["NL_NL_EMMAVOICE"] = "nl-NL_EmmaVoice";
            Voice["NL_NL_LIAMVOICE"] = "nl-NL_LiamVoice";
            Voice["PT_BR_ISABELAVOICE"] = "pt-BR_IsabelaVoice";
            Voice["PT_BR_ISABELAV3VOICE"] = "pt-BR_IsabelaV3Voice";
            Voice["ZH_CN_LINAVOICE"] = "zh-CN_LiNaVoice";
            Voice["ZH_CN_WANGWEIVOICE"] = "zh-CN_WangWeiVoice";
            Voice["ZH_CN_ZHANGJINGVOICE"] = "zh-CN_ZhangJingVoice";
        })(Voice = SynthesizeConstants.Voice || (SynthesizeConstants.Voice = {}));
    })(SynthesizeConstants = TextToSpeechV1.SynthesizeConstants || (TextToSpeechV1.SynthesizeConstants = {}));
    /** Constants for the `getPronunciation` operation. */
    var GetPronunciationConstants;
    (function (GetPronunciationConstants) {
        /** A voice that specifies the language in which the pronunciation is to be returned. All voices for the same language (for example, `en-US`) return the same translation. For more information about specifying a voice, see **Important voice updates** in the method description. */
        var Voice;
        (function (Voice) {
            Voice["AR_AR_OMARVOICE"] = "ar-AR_OmarVoice";
            Voice["AR_MS_OMARVOICE"] = "ar-MS_OmarVoice";
            Voice["DE_DE_BIRGITVOICE"] = "de-DE_BirgitVoice";
            Voice["DE_DE_BIRGITV3VOICE"] = "de-DE_BirgitV3Voice";
            Voice["DE_DE_DIETERVOICE"] = "de-DE_DieterVoice";
            Voice["DE_DE_DIETERV3VOICE"] = "de-DE_DieterV3Voice";
            Voice["DE_DE_ERIKAV3VOICE"] = "de-DE_ErikaV3Voice";
            Voice["EN_AU_CRAIGVOICE"] = "en-AU-CraigVoice";
            Voice["EN_AU_MADISONVOICE"] = "en-AU-MadisonVoice";
            Voice["EN_GB_CHARLOTTEV3VOICE"] = "en-GB_CharlotteV3Voice";
            Voice["EN_GB_JAMESV3VOICE"] = "en-GB_JamesV3Voice";
            Voice["EN_GB_KATEVOICE"] = "en-GB_KateVoice";
            Voice["EN_GB_KATEV3VOICE"] = "en-GB_KateV3Voice";
            Voice["EN_US_ALLISONVOICE"] = "en-US_AllisonVoice";
            Voice["EN_US_ALLISONV3VOICE"] = "en-US_AllisonV3Voice";
            Voice["EN_US_EMILYV3VOICE"] = "en-US_EmilyV3Voice";
            Voice["EN_US_HENRYV3VOICE"] = "en-US_HenryV3Voice";
            Voice["EN_US_KEVINV3VOICE"] = "en-US_KevinV3Voice";
            Voice["EN_US_LISAVOICE"] = "en-US_LisaVoice";
            Voice["EN_US_LISAV3VOICE"] = "en-US_LisaV3Voice";
            Voice["EN_US_MICHAELVOICE"] = "en-US_MichaelVoice";
            Voice["EN_US_MICHAELV3VOICE"] = "en-US_MichaelV3Voice";
            Voice["EN_US_OLIVIAV3VOICE"] = "en-US_OliviaV3Voice";
            Voice["ES_ES_ENRIQUEVOICE"] = "es-ES_EnriqueVoice";
            Voice["ES_ES_ENRIQUEV3VOICE"] = "es-ES_EnriqueV3Voice";
            Voice["ES_ES_LAURAVOICE"] = "es-ES_LauraVoice";
            Voice["ES_ES_LAURAV3VOICE"] = "es-ES_LauraV3Voice";
            Voice["ES_LA_SOFIAVOICE"] = "es-LA_SofiaVoice";
            Voice["ES_LA_SOFIAV3VOICE"] = "es-LA_SofiaV3Voice";
            Voice["ES_US_SOFIAVOICE"] = "es-US_SofiaVoice";
            Voice["ES_US_SOFIAV3VOICE"] = "es-US_SofiaV3Voice";
            Voice["FR_FR_NICOLASV3VOICE"] = "fr-FR_NicolasV3Voice";
            Voice["FR_FR_RENEEVOICE"] = "fr-FR_ReneeVoice";
            Voice["FR_FR_RENEEV3VOICE"] = "fr-FR_ReneeV3Voice";
            Voice["IT_IT_FRANCESCAVOICE"] = "it-IT_FrancescaVoice";
            Voice["IT_IT_FRANCESCAV3VOICE"] = "it-IT_FrancescaV3Voice";
            Voice["JA_JP_EMIVOICE"] = "ja-JP_EmiVoice";
            Voice["JA_JP_EMIV3VOICE"] = "ja-JP_EmiV3Voice";
            Voice["KO_KR_HYUNJUNVOICE"] = "ko-KR_HyunjunVoice";
            Voice["KO_KR_SIWOOVOICE"] = "ko-KR_SiWooVoice";
            Voice["KO_KR_YOUNGMIVOICE"] = "ko-KR_YoungmiVoice";
            Voice["KO_KR_YUNAVOICE"] = "ko-KR_YunaVoice";
            Voice["NL_NL_EMMAVOICE"] = "nl-NL_EmmaVoice";
            Voice["NL_NL_LIAMVOICE"] = "nl-NL_LiamVoice";
            Voice["PT_BR_ISABELAVOICE"] = "pt-BR_IsabelaVoice";
            Voice["PT_BR_ISABELAV3VOICE"] = "pt-BR_IsabelaV3Voice";
            Voice["ZH_CN_LINAVOICE"] = "zh-CN_LiNaVoice";
            Voice["ZH_CN_WANGWEIVOICE"] = "zh-CN_WangWeiVoice";
            Voice["ZH_CN_ZHANGJINGVOICE"] = "zh-CN_ZhangJingVoice";
        })(Voice = GetPronunciationConstants.Voice || (GetPronunciationConstants.Voice = {}));
        /** The phoneme format in which to return the pronunciation. The Arabic, Chinese, Dutch, Australian English, and Korean languages support only IPA. Omit the parameter to obtain the pronunciation in the default format. */
        var Format;
        (function (Format) {
            Format["IBM"] = "ibm";
            Format["IPA"] = "ipa";
        })(Format = GetPronunciationConstants.Format || (GetPronunciationConstants.Format = {}));
    })(GetPronunciationConstants = TextToSpeechV1.GetPronunciationConstants || (TextToSpeechV1.GetPronunciationConstants = {}));
    /** Constants for the `createCustomModel` operation. */
    var CreateCustomModelConstants;
    (function (CreateCustomModelConstants) {
        /** The language of the new custom model. You create a custom model for a specific language, not for a specific voice. A custom model can be used with any voice for its specified language. Omit the parameter to use the the default language, `en-US`. **Note:** The `ar-AR` language identifier cannot be used to create a custom model. Use the `ar-MS` identifier instead. */
        var Language;
        (function (Language) {
            Language["AR_MS"] = "ar-MS";
            Language["DE_DE"] = "de-DE";
            Language["EN_GB"] = "en-GB";
            Language["EN_US"] = "en-US";
            Language["ES_ES"] = "es-ES";
            Language["ES_LA"] = "es-LA";
            Language["ES_US"] = "es-US";
            Language["FR_FR"] = "fr-FR";
            Language["IT_IT"] = "it-IT";
            Language["JA_JP"] = "ja-JP";
            Language["KO_KR"] = "ko-KR";
            Language["NL_NL"] = "nl-NL";
            Language["PT_BR"] = "pt-BR";
            Language["ZH_CN"] = "zh-CN";
        })(Language = CreateCustomModelConstants.Language || (CreateCustomModelConstants.Language = {}));
    })(CreateCustomModelConstants = TextToSpeechV1.CreateCustomModelConstants || (TextToSpeechV1.CreateCustomModelConstants = {}));
    /** Constants for the `listCustomModels` operation. */
    var ListCustomModelsConstants;
    (function (ListCustomModelsConstants) {
        /** The language for which custom models that are owned by the requesting credentials are to be returned. Omit the parameter to see all custom models that are owned by the requester. */
        var Language;
        (function (Language) {
            Language["AR_MS"] = "ar-MS";
            Language["DE_DE"] = "de-DE";
            Language["EN_AU"] = "en-AU";
            Language["EN_GB"] = "en-GB";
            Language["EN_US"] = "en-US";
            Language["ES_ES"] = "es-ES";
            Language["ES_LA"] = "es-LA";
            Language["ES_US"] = "es-US";
            Language["FR_FR"] = "fr-FR";
            Language["IT_IT"] = "it-IT";
            Language["JA_JP"] = "ja-JP";
            Language["KO_KR"] = "ko-KR";
            Language["NL_NL"] = "nl-NL";
            Language["PT_BR"] = "pt-BR";
            Language["ZH_CN"] = "zh-CN";
        })(Language = ListCustomModelsConstants.Language || (ListCustomModelsConstants.Language = {}));
    })(ListCustomModelsConstants = TextToSpeechV1.ListCustomModelsConstants || (TextToSpeechV1.ListCustomModelsConstants = {}));
    /** Constants for the `addWord` operation. */
    var AddWordConstants;
    (function (AddWordConstants) {
        /** **Japanese only.** The part of speech for the word. The service uses the value to produce the correct intonation for the word. You can create only a single entry, with or without a single part of speech, for any word; you cannot create multiple entries with different parts of speech for the same word. For more information, see [Working with Japanese entries](https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-rules#jaNotes). */
        var PartOfSpeech;
        (function (PartOfSpeech) {
            PartOfSpeech["DOSI"] = "Dosi";
            PartOfSpeech["FUKU"] = "Fuku";
            PartOfSpeech["GOBI"] = "Gobi";
            PartOfSpeech["HOKA"] = "Hoka";
            PartOfSpeech["JODO"] = "Jodo";
            PartOfSpeech["JOSI"] = "Josi";
            PartOfSpeech["KATO"] = "Kato";
            PartOfSpeech["KEDO"] = "Kedo";
            PartOfSpeech["KEYO"] = "Keyo";
            PartOfSpeech["KIGO"] = "Kigo";
            PartOfSpeech["KOYU"] = "Koyu";
            PartOfSpeech["MESI"] = "Mesi";
            PartOfSpeech["RETA"] = "Reta";
            PartOfSpeech["STBI"] = "Stbi";
            PartOfSpeech["STTO"] = "Stto";
            PartOfSpeech["STZO"] = "Stzo";
            PartOfSpeech["SUJI"] = "Suji";
        })(PartOfSpeech = AddWordConstants.PartOfSpeech || (AddWordConstants.PartOfSpeech = {}));
    })(AddWordConstants = TextToSpeechV1.AddWordConstants || (TextToSpeechV1.AddWordConstants = {}));
})(TextToSpeechV1 || (TextToSpeechV1 = {}));
module.exports = TextToSpeechV1;
